core/src/main/java/org/apache/iceberg/LocationProviders.java

```diff
@@ -172,10 +193,45 @@ private static String pathContext(String tableLocation) {
     }
 
     private String computeHash(String fileName) {
-      byte[] bytes = TEMP.get();
-      HashCode hash = HASH_FUNC.hashString(fileName, StandardCharsets.UTF_8);
-      hash.writeBytesTo(bytes, 0, 4);
-      return BASE64_ENCODER.encode(bytes);
+      HashCode hashCode = HASH_FUNC.hashString(fileName, StandardCharsets.UTF_8);
+
+      // {@link Integer#toBinaryString} excludes leading zeros, which we want to preserve.
+      // force the first bit to be set to get around that.
+      String hashAsBinaryString = Integer.toBinaryString(hashCode.asInt() | Integer.MIN_VALUE);
+      return dirsFromHash(hashAsBinaryString.substring(HASH_BINARY_STRING_START_INDEX));
+    }
+
+    /**
+     * Divides hash into directories for optimized orphan removal operation using ENTROPY_DIR_DEPTH
+     * and ENTROPY_DIR_LENGTH
+     *
+     * @param hash 10011001100110011001
+     * @return 1001/1001/1001/10011001 with depth 3 and length 4
+     */
+    private String dirsFromHash(String hash) {
+      if (hash.length() < ENTROPY_DIR_DEPTH * ENTROPY_DIR_LENGTH) {
```

Comment:
We should use a `Preconditions.check` here.

Reply:
Yeah maybe I am a bit over-paranoid here, given I have the right unit tests to cover any changes and length is 20 bits as final string(after addressing the other feedback), maybe will just remove this condition?

Reply:
If it's all internal to this class, I think we can safely assume that `Integer.toBinaryString(hashCode.asInt() | Integer.MIN_VALUE)` will produce the correct number of bits.  The check is seems unnecessary since it's not user supplied inputs.

Reply:
Done